src/raster/buffer.rs

```diff
@@ -1,40 +1,114 @@
-    pub data: Vec<T>,
+    shape: (usize, usize),
+    data: Vec<T>,
 }
 
 impl<T: GdalType> Buffer<T> {
     /// Construct a new buffer from `size` (`(cols, rows)`) and `Vec<T>`.
     ///
-    /// # Panic
+    /// # Panics
     /// Will panic if `size.0 * size.1 != data.len()`.
-    pub fn new(size: (usize, usize), data: Vec<T>) -> Self {
+    pub fn new(shape: (usize, usize), data: Vec<T>) -> Self {
         assert_eq!(
-            size.0 * size.1,
+            shape.0 * shape.1,
             data.len(),
             "size {:?} does not match length {}",
-            size,
+            shape,
             data.len()
         );
-        Buffer { size, data }
+        Buffer { shape, data }
+    }
+
+    /// Destructures `self` into constituent parts.
+    pub fn into_shape_and_vec(self) -> ((usize, usize), Vec<T>) {
+        (self.shape, self.data)
+    }
+
+    /// Gets the 2-d shape of the buffer.
+    ///
+    /// Returns `(cols, rows)`
+    pub fn shape(&self) -> (usize, usize) {
+        self.shape
+    }
+
+    /// Get a slice over the buffer contents.
+    pub fn data(&self) -> &[T] {
+        self.data.as_slice()
+    }
+
+    /// Get a mutable slice over the buffer contents.
+    pub fn data_mut(&mut self) -> &mut [T] {
+        self.data.as_mut_slice()
+    }
+
+    /// Get the number of elements in the buffer
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Determine if the buffer has no elements.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
     }
 
     #[cfg(feature = "ndarray")]
     /// Convert `self` into an [`ndarray::Array2`].
     pub fn to_array(self) -> crate::errors::Result<Array2<T>> {
         // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
         Ok(Array2::from_shape_vec(
-            (self.size.1, self.size.0),
+            (self.shape.1, self.shape.0),
             self.data,
         )?)
     }
+
+    #[inline]
+    pub(self) fn vec_index_for(&self, coord: (usize, usize)) -> usize {
+        if coord.0 >= self.shape.0 {
+            panic!(
+                "index out of bounds: buffer has {} columns but row {} was requested",
+                self.shape.0, coord.0
+            );
+        }
+        if coord.1 >= self.shape.1 {
+            panic!(
+                "index out of bounds: buffer has {} rows but row {} was requested",
+                self.shape.1, coord.1
+            );
+        }
+        coord.0 * self.shape.0 + coord.1
+    }
+}
+
+impl<T: GdalType> Index<(usize, usize)> for Buffer<T> {
+    type Output = T;
+    fn index(&self, index: (usize, usize)) -> &Self::Output {
+        &self.data[self.vec_index_for(index)]
+    }
+}
+
+impl<T: GdalType> IndexMut<(usize, usize)> for Buffer<T> {
+    fn index_mut(&mut self, index: (usize, usize)) -> &mut Self::Output {
+        let idx = self.vec_index_for(index);
+        &mut self.data[idx]
+    }
+}
+
+impl<'a, T: GdalType> IntoIterator for &'a Buffer<T> {
```

Comment:
We want three implementations, for `Buffer<T>`, for `&Buffer<T>` and for `&mut Buffer<T>`.